components/hab/src/main.rs

```diff
@@ -1676,6 +1686,109 @@ fn idents_from_file_matches(matches: &ArgMatches<'_>) -> Result<Vec<PackageIdent
     Ok(sources)
 }
 
+fn idents_from_toml_matches(matches: &ArgMatches<'_>) -> Result<Vec<PackageSet>> {
```

Comment:
Somewhat concerned that we have to do some much manual construction of the package set from the TOML...

Reply:
I haven't dug into this PR yet (this is more of a drive-by comment at this point), but I notice that there isn't an implementation of `Deserialize` for `PackageSet`. Having that would probably simplify a fair bit of this logic, if not remove it altogether.

Reply:
Second, simplifying with `Deserialize`. Here is a stab at the types:

```rust
#[derive(Debug, Deserialize)]
pub struct PackageSetFile {
    pub format_version:  u8,
    pub file_descriptor: String,
    pub package_sets:    HashMap<PackageTarget, Vec<PackageSetValue>>,
}

#[derive(Debug, Deserialize)]
pub struct PackageSetValue {
    pub channel: ChannelIdent,
    pub idents:  Vec<PackageIdent>,
}

#[derive(Debug, Deserialize)]
pub struct PackageSet {
    pub target:  PackageTarget,
    pub channel: ChannelIdent,
    pub idents:  Vec<PackageIdent>,
}
```

This will lead to needing a `package_sets` key so the file format will need to use something like `[["package_sets.x86_64-linux"]]`. The other option would be to use some macro magic to add a key for each `PackageTarget` variant

Reply:
@davidMcneil I did try something like the above, but hit some snags around PackageTarget being a key, and the already defined serializer for PackageIdent. Working on some ideas now.

Reply:
@markan serde's field attribute [`#[serde(deserialize_with = "path")]`](https://serde.rs/field-attrs.html#deserialize_with) might help if we need a custom deserializer?

Reply:
Implemented